### PR TITLE
Apply builder pattern to kernel launcher

### PR DESCRIPTION
**Pull request by ZzEeKkAa**
Update kernel launcher builder to use builder pattern. Since the argument list is long we need modular way of providing those arguments to make code more readable and extendable.

- [x] Have you provided a meaningful PR description?
- [ ] Have you added a test, reproducer or referred to an issue with a reproducer?
- [x] Have you tested your changes locally for CPU and GPU devices?
- [x] Have you made sure that new changes do not introduce compiler warnings?
- [ ] If this PR is a work in progress, are you filing the PR as a draft?

### CODE REVIEW

File: numba_dpex/core/utils/kernel_launcher.py [hunk 1]
```diff
@@ -68,15 +116,15 @@ def _build_array_attr_arg(
...
+        self._build_arg(
             val=array_attr,
             ty=array_attr_ty,
             arg_list=arg_list,
             args_ty_list=args_ty_list,
             arg_num=arg_num,
         )
 
-    def _build_unituple_member_arg(
+    def _build_unituple_member_arg(  # pylint: disable=too-many-arguments
```
Comment (diptorupd) [inline numba_dpex/core/utils/kernel_launcher.py:127]: Adding the pylint disabling is too much noise in code. Following linter guide line is OK, but there should be room to tweak.
For this particular check , can you update the `pylintrc` max-args to 6.

File: numba_dpex/core/parfors/parfor_lowerer.py [hunk 2]
```diff
@@ -68,11 +69,8 @@ def _getvar(lowerer, x):
         var_val = lowerer.varmap[x]
 
     if var_val:
-        if not isinstance(var_val.type, llvmir.PointerType):
-            with lowerer.builder.goto_entry_block():
-                var_val_ptr = lowerer.builder.alloca(var_val.type)
-            lowerer.builder.store(var_val, var_val_ptr)
-            return var_val_ptr
+        if isinstance(var_val.type, llvmir.PointerType):
```
Comment (diptorupd) [inline numba_dpex/core/parfors/parfor_lowerer.py:72]: Why is this change needed? Is `var_val` always a `PointerType`?

File: numba_dpex/core/utils/kernel_launcher.py (hunk 1 above)
Comment (ZzEeKkAa) [reply numba_dpex/core/utils/kernel_launcher.py:127]: I don't want to apply global settings for the number of max-args cause it affects readability. It is a todo work to update this code to use less arguments.

File: numba_dpex/core/parfors/parfor_lowerer.py (hunk 2 above)
Comment (ZzEeKkAa) [reply numba_dpex/core/parfors/parfor_lowerer.py:72]: It used to return pointer to value, not it returns value as description says.

Comment (ZzEeKkAa) [reply numba_dpex/core/parfors/parfor_lowerer.py:72]: Change needed in parfor to be aligned with the semantic of numba's unboxed llvm values. Resulted llvmir is not affected, since value to ptr concertation is done in another step ( `KernelLauncerIRBuilder.set_arguments` ).

Comment (ZzEeKkAa) [reply numba_dpex/core/parfors/parfor_lowerer.py:72]: It also makes sense because `ParforKernel.kernel_args_ty` now actually match types of the llvm values.

File: numba_dpex/core/parfors/parfor_lowerer.py [hunk 3]
```diff
@@ -19,10 +19,11 @@
     ReductionKernelVariables,
 )
 from numba_dpex.core.utils.kernel_launcher import KernelLaunchIRBuilder
+from numba_dpex.dpctl_iface import libsyclinterface_bindings as sycl
```
Comment (diptorupd) [inline numba_dpex/core/parfors/parfor_lowerer.py:22]: `from numba_dpex.core.datamodel.models import dpex_data_model_manager as dpex_dmm` is now unused please remove as part of the PR.

Comment (ZzEeKkAa) [reply numba_dpex/core/parfors/parfor_lowerer.py:22]: And `_KernelArgs`

Comment (diptorupd) [reply numba_dpex/core/parfors/parfor_lowerer.py:21]: yes

**Comment by diptorupd** [comment]
Few dead code that I noticed after your changes in `parfor_lowerer.py`

- [ ] `_KernelArgs`
- [ ] from numba_dpex.core.datamodel.models import dpex_data_model_manager as dpex_dmm

File: numba_dpex/core/parfors/parfor_lowerer.py [hunk 4]
```diff
@@ -176,45 +108,13 @@ def _submit_parfor_kernel(
 
         local_range = []
```
Comment (diptorupd) [inline numba_dpex/core/parfors/parfor_lowerer.py:109]: A comment is warranted here: "For now the `local_range` is always an empty list as numba_dpex always submits kernels generated for parfor nodes as `range` kernels. The provision is kept here if in future there is newer functionality to submit these kernels as `ndrange`."

File: numba_dpex/core/parfors/parfor_lowerer.py [hunk 5]
```diff
@@ -228,75 +128,61 @@ def _submit_reduction_main_parfor_kernel(
...
-            global_range=global_range,
-            local_range=local_range,
+
+        kl_builder.set_kernel(kernel_ref)
+        kl_builder.set_queue(queue_ref)
+        kl_builder.set_range(global_range, local_range)
+        kl_builder.set_arguments(
+            kernel_fn.kernel_arg_types, kernel_args=kernel_args
         )
+        kl_builder.set_dependant_event_list()
```
Comment (diptorupd) [inline numba_dpex/core/parfors/parfor_lowerer.py:180]: A comment is required: "Kernels generated for `parfor` nodes are always submitted synchronously and do not allow for any dependent events. Thus, the `kl_builder.set_dependant_event_list()` is called without any arguments."

I would have preferred the API forces you to pass in an empty list/tuple. Otherwise the call here can be confusing.

File: numba_dpex/core/parfors/reduction_helper.py [hunk 6]
```diff
@@ -399,7 +399,7 @@ def copy_final_sum_to_host(self, parfor_kernel):
 
         # Create a local variable storing a pointer to a DPCTLSyclQueueRef
         # pointer.
-        curr_queue = ir_builder.get_queue(exec_queue=parfor_kernel.queue)
+        queue_ref = ir_builder.get_queue(exec_queue=parfor_kernel.queue)
```
Comment (diptorupd) [inline numba_dpex/core/parfors/reduction_helper.py:402]: Let us stay consistent with naming. `s/ir\_builder/kl\_builder/g`

File: numba_dpex/core/utils/kernel_launcher.py [hunk 7]
```diff
@@ -151,11 +204,11 @@ def build_complex_arg(self, val, ty, arg_list, args_ty_list, arg_num):
         )
         arg_num += 1
 
-    def build_array_arg(
+    def _build_array_arg(  # pylint: disable=too-many-arguments
         self,
         array_val,
         array_data_model,
-        array_rank,
+        array_rank,  # pylint: disable=unused-argument
```
Comment (diptorupd) [inline numba_dpex/core/utils/kernel_launcher.py:211]: Why not remove?

File: numba_dpex/core/parfors/reduction_helper.py (hunk 6 above)
Comment (diptorupd) [reply numba_dpex/core/parfors/reduction_helper.py:402]: Also, I am assuming the `get_queue` will itself be cleaned up in a follow up PR?

File: numba_dpex/experimental/launcher.py [hunk 8]
```diff
@@ -317,16 +317,15 @@ def codegen(cgctx, builder, sig, llargs):
...
-            local_range_extents=ll_range.local_range_extents,
+        kl_builder = kl.KernelLaunchIRBuilder(kernel_targetctx, builder)
+        kl_builder.set_kernel(kernel_ref)
+        kl_builder.set_queue(queue_ref)
+        kl_builder.set_range(
+            ll_range.global_range_extents, ll_range.local_range_extents
         )
+        kl_builder.set_arguments(ty_kernel_args_tuple, kernel_args)
+        kl_builder.set_dependant_event_list()
+        device_event_ref = kl_builder.submit()
```
Comment (diptorupd) [inline numba_dpex/experimental/launcher.py:328]: I like the pattern. Much cleaner.

File: numba_dpex/core/utils/kernel_launcher.py [hunk 9]
```diff
@@ -11,6 +17,48 @@
 from numba_dpex.dpctl_iface import libsyclinterface_bindings as sycl
 from numba_dpex.dpctl_iface._helpers import numba_type_to_dpctl_typenum
 
+MAX_SIZE_OF_SYCL_RANGE = 3
```
Comment (diptorupd) [inline numba_dpex/core/utils/kernel_launcher.py:20]: There is a place in `parfor_lowerer.py` where the same hard coing is done. It will be goo to move the global to `dpctl_iface` and use it from there.

**Review by diptorupd — APPROVED** [review]
Nice clean up. Makes the code readable and reduces redundancies.

**Comment by ZzEeKkAa** [comment]
There is an issue in this change. Before this change parfor_lowerer was using `from numba_dpex.core.datamodel.models import dpex_data_model_manager as dpex_dmm` data model manager, which is the same as kernel data model manager, but now it uses data model manager from context which is equal to numba's data model manager. It is okay as long as number of arguments for USMNdArray and DpnpNdArray is the same, but I would introduce KernelLaunchIRBuilder.kernel_data_model_manager attribute to store the right dmm on instance creation.

File: numba_dpex/core/utils/kernel_launcher.py (hunk 7 above)
Comment (ZzEeKkAa) [reply numba_dpex/core/utils/kernel_launcher.py:211]: I've thought it was pending TODO. Removed

File: numba_dpex/core/parfors/reduction_helper.py (hunk 6 above)
Comment (ZzEeKkAa) [reply numba_dpex/core/parfors/reduction_helper.py]: Yes, it is not part of builder, so it's gonna be removed. Either by getting queue on llvm level or removing this code out of the class.

File: numba_dpex/core/parfors/parfor_lowerer.py (hunk 5 above)
Comment (ZzEeKkAa) [reply numba_dpex/core/parfors/parfor_lowerer.py:188]: Good idea!
